dvc/repo/metrics/show.py

```diff
@@ -22,17 +22,29 @@ def _collect_metrics(repo, targets, recursive):
 
     if not targets:
         return list(metrics)
-
-    target_infos = [PathInfo(os.path.abspath(target)) for target in targets]
-
-    def _filter(path_info):
-        for info in target_infos:
-            func = path_info.isin_or_eq if recursive else path_info.__eq__
-            if func(info):
-                return True
-        return False
-
-    return list(filter(_filter, metrics))
+    else:
+        result = []
+        if recursive:
+            target_infos = [
+                PathInfo(os.path.abspath(target)) for target in targets
+            ]
+
+            def _filter(path_info):
+                for info in target_infos:
+                    func = (
+                        path_info.isin_or_eq if recursive else path_info.__eq__
+                    )
+                    if func(info):
+                        return True
+                return False
+
+            result.extend(list(filter(_filter, metrics)))
+
+        tree = RepoTree(repo)
+        for target in targets:
+            if tree.exists(target) and not tree.isdir(target):
```

Comment:
```suggestion
            if tree.isfile(target):
```
?

Reply:
Not strictly equivalent, but wonder if that's what you meant here.

Reply:
Well I wanted to omit directories in case of `recursive`, so applying some logic `isfile` seems like a better check.